### DISCUSSION TITLE
Is it possible to override template files for plugins?

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3424] [Q&A] [answered]
I'm using `neoteroi.mkdocsoad` to generate docs from Open API specifications... And after looking a bit into it did I notice that the tool it uses - essentials-openapi - apparently utilizes Jina2 templating for MkDocs-style documentation.

I now wonder: Can I somehow override the used template files of the plugin to customize some of its output?
If I can, are there ways to figure out the effective path of a template file, so that I could override them?

I also asked in the essentials-openapi repository, but given the maintainer's recent activity here on GitHub I have doubts to receive any reply to it...
Worst case I have to create a forked version of the plugin to modify it... Which opens a whole can of worms I would love to keep closed as I'm not ready yet to tackle Python and MkDocs plugins...

One alternative I've been thinking about here is to maybe use templating itself with the openapi file? Would that be a thing to work with?

**Comment by oprypin** [comment, marked as answer]
This question is 100% specific to the plugin. But just pointing that out, it's not forbidden to ask.

essentials-openapi seems pluggable, you'd just provide your own package with themes.
https://github.com/Neoteroi/essentials-openapi/blob/6bbd8c51934c65ec2c315797f6fb4b64bba14dc8/openapidocs/mk/v3/__init__.py#L101
https://github.com/Neoteroi/essentials-openapi/blob/6bbd8c51934c65ec2c315797f6fb4b64bba14dc8/openapidocs/mk/jinja.py#L101

The plugin does not seem to expose this pluggability, but on the other hand the plugin itself is trivial, it's just this 1 file 
https://github.com/Neoteroi/mkdocs-plugins/blob/d196ab67594e9cd6adcdda75089bd7020253a08e/neoteroi/mkdocs/oad/__init__.py#L32

**Reply by oprypin** [reply]
(so you could potentially make a copy of the plugin and a copy of the templates and change them how you like, with the only dependency then being essentials-openapi)

**Reply by Andre601** [reply]
I was wondering if this is even worth it?

Like... The source itself is a JSON file... So would it be doable to pull the info and parse it with my own template files? I assume to may need some python hook to work? If it works that is...

**Reply by oprypin** [reply]
Seems like a lot less work to reuse their setup. 

Just copy their plugin (it's just a single file) (possibly as a hook instead, dropping `self, ` in that case)

but replace this part
https://github.com/Neoteroi/mkdocs-plugins/blob/d196ab67594e9cd6adcdda75089bd7020253a08e/neoteroi/mkdocs/oad/__init__.py#L31-L33
with the following:
```python
OpenAPIV3DocumentationHandler(
    data, source=source, write=openapidocs.mk.jinja.Jinja2DocumentsWriter('my_oad_templates')
)
```

Copy this folder https://github.com/Neoteroi/essentials-openapi/tree/main/openapidocs/mk/v3/views_mkdocs
as 'my_oad_templates/views_mkdocs'
and edit it as you like

**Reply by oprypin** [reply]
* You'll most likely need the workaround for https://github.com/mkdocs/mkdocs/issues/3281

**Reply by Andre601** [reply]
Hey. Coming back to this as I now actually need to do this...

I'm not entirely sure if I understand you correctly... The plugin itself has `class MkDocsOpenAPIDocumentationPlugin(BasePlugin):` defined at the top. Would I need to include this, if I add my own version as a hook, or can this be droped?

Also, by droping the self part, do you mean having the file setup like this?
```python
import re

from mkdocs.config.config_options import Type
from mkdocs.plugins import BasePlugin
from openapidocs.mk.v3 import OpenAPIV3DocumentationHandler
from openapidocs.utils.source import read_from_source

rx = re.compile(r"\[OAD\(([^\)]+)\)\]")

def _replacer(match) -> str:
    source = match.group(1).strip("'\"")
    data = read_from_source(source)
    
    handler = OpenAPIV3DocumentationHandler(
        data, source=source, writer=openapidocs.mk.jinja.Jinja2DocumentsWriter('my_oad_template')
    )

def on_page_markdown(markdown: str, *args, **kwargs):
    if "[ODAD(" in markdown:
        return rx.sub(_replacer, markdown)
    return markdown
```
On another note, VSCode reports t me that `openapidocs` is not defined in the `writer=` part (btw, I think your example has a type as VSCode only suggests writer here and not write)

**Reply by Andre601** [reply]
Also, would the path for the template be relative to the mkdocs config file, the hook, or something else?

I need to know so that I can set a proper path to use here (I also assume the path uses the `/` path divider?)

**Reply by Andre601** [reply]
I fixed the jinja thing by having a `from openapidocs.mk.jinja import Jinja2DocumentsWriter` line at the top.

Tho, right now can it not find the `theme` module I defined... Is that even the right thing to give? Because looking at essentials-openapi, it seems that `Jinja2DocumentsWriter` wants a `DocumentsWriter`: https://github.com/Neoteroi/essentials-openapi/blob/main/openapidocs/mk/jinja.py#L90

(Also, yes I changed `my_oad_template` to `theme` in the code.)

**Reply by oprypin** [reply]
> would the path for the template be relative to the mkdocs config file, the hook, or something else?

Because `Jinja2DocumentsWriter` uses https://jinja.palletsprojects.com/en/3.0.x/api/#jinja2.PackageLoader, the path is relative to Python's import path. That's why if the hook script and the template directory are in the same common directory, it is good to write `sys.path.insert(0, os.path.dirname(__file__))` to put the common directory into Python's import path.

**Reply by Andre601** [reply]
That works. Thank you. Now... time to edit this stuff xD

**Reply by oprypin** [reply]
Great! And yes your code looks good, it is exactly what I meant

**Reply by Andre601** [reply]
Looks good!

![grafik](https://github.com/mkdocs/mkdocs/assets/11576465/303af45e-2e3d-4537-b93d-2fd9d6da447e)
